apf-ingester/apf_Ingester.py

```diff
@@ -0,0 +1,264 @@
+            raise ValueError("Only one of apf path and apf string may be specified")
+        if not apf_path and not apf_string:
+            raise ValueError("One of apf_path or apf_string must be specified")
+
+        if apf_string:
+            self.apf_string = apf_string
+        else:
+            with open(apf_path, 'r') as apf_file:
+                self.apf_string = apf_file.read()
+
+    def extract_event_list_to_dict(self,
+                                   current_project_dict: Dict[str, List[str]]) -> Dict[str,
+                                                                                       List[str]]:
+        """Extracts events from the apf file, adding it to the current_project_dict. Utilizes a
+        cache to prevent duplicates
+
+        Args:
+            current_project_dict (Dict[str, List[str]]): Existing project dict containing other
+            Event to Document mappings.
+
+        Returns:
+            A Dictionary mapping EVENT_TYPE.EVENT_SUBTYPE to a List of DOC-IDs
+
+        """
+        # Cache to prevent duplicate entries (in case there are multiple of the same event in a
+        # document)
+        visited_keys = []
+        for event in re.findall(r"<event ID=.*?>", self.apf_string):
+            matcher = re.fullmatch(r'<event ID="(.*?)-EV(.*?)" TYPE="(.*?)" SUBTYPE="(.*?)".*?>',
+                                   event)
+            if matcher is not None:
+                key_name: str = matcher.group(3) + "." + matcher.group(4)
+                doc_id: str = matcher.group(1)
+                if key_name in visited_keys:
+                    continue
+                if key_name in current_project_dict:
+                    current_project_dict.get(key_name).append(doc_id)
+                else:
+                    current_project_dict[key_name] = [doc_id]
+                visited_keys.append(key_name)
+        return current_project_dict
+
+
+def get_complete_project_to_doc_mapping(ace_corpus_path: str) -> Dict[str, List[str]]:
+    """
+
+    Args:
+        ace_corpus_path: Path to the corpus directory containing the .sgm and .apf files
+
+    Returns:
+         Dict[str, List[str]]: Dict mapping EVENT_TYPE.EVENT_SUBTYPE to a list of document
+    names that contain a matching event
+
+    """
+    complete_project_map: Dict[str, List[str]] = {}
+    for filename in os.listdir(ace_corpus_path):
+        if filename.endswith(".sgm"):
+            filename_without_extension = filename.split(".sgm")[0]
+            source_path = ace_corpus_path + filename
+            apf_path = ace_corpus_path + filename_without_extension + ".apf.xml"
+            ingester_input = ApfIngesterInput(apf_path)
+            ingester_input.extract_event_list_to_dict(complete_project_map)
+    # Add empty entry for 'None' type, to generate an empty project
+    complete_project_map["None"] = []
+    return complete_project_map
+
+
+def configure_and_generate_project(json_template_path: str,

```

Comment:
When you have a lot of arguments of similar type, prefer to make them keyword only (in this case, place a `*` first in the argument list, like `(*, `.